### ISSUE TITLE
Add MPN to products table and

### ISSUE THREAD

**Issue by torvista** [zencart/zencart#5727]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**
Important: Explain what "Business Problem" is being solved by this solution. 
What kind of merchant (small business, retail-only, digital-products-only, etc) would most benefit from this solution. Who might "not" benefit much?

**Describe alternatives you've considered**
If you have working code that could be a possible solution, include a link to where that code can be viewed.

**Additional context**
Add any other context or screenshots about the feature request here. Include URLs where possible solutions can be seen, if any.

**Comment by lat9** [comment]
Good idea, but since it changes the database schema, it'll need to be deferred until the next unsuffixed version.

**Comment by scottcwilson** [comment]
I want to include this in the next release.

**Comment by lat9** [comment]
Noting also that it makes no sense to add an MPN to a `products_attributes` record.  An MPN for an attributed product would apply to the product+_**attribute-combinations**_, e.g. shirt+size+color,  Having a value in each attributes' record is a non-starter.

**Comment by drbyte** [comment]
Remember that you may want the `orders_products` table to also store the information as it was at the time of an order.

**Comment by lat9** [comment]
Does the inclusion of `products_mpn` necessitate the addition of a "SHOW_PRODUCT{_type}_INFO_MPN" setting in the `product_type_layout` table and related changes to the various `tpl_product{_type}_info_display.php` templates?

![image](https://github.com/zencart/zencart/assets/2685585/4046da8e-c114-47fc-9fb8-7cc1bd974018)

**Comment by scottcwilson** [comment]
I would vote no.  People who want it can just update their templates, which they'll have to do anyhow.

**Comment by lat9** [comment]
How about the function addition?  Noting that any database update would start with the `products_mpn` value as `NULL` to distinguish 'unset' vs. not supplied.

**Comment by scottcwilson** [comment]
Sure, function is fine.

**Comment by lat9** [comment]
Sigh, that addition will also necessitate updates to the admin's various product types' `collect_info.php` modules as well as the `update_product.php` module.

**Comment by scottcwilson** [comment]
Yep.

**Comment by proseLA** [comment]
> Remember that you may want the `orders_products` table to also store the information as it was at the time of an order.

i disagree.  MPNs are assigned by the manufacturer, and therefore rarely change for a specific product.  and if there is a change, in theory it is a different product.  ensuring that you have the MPN at the time of the order in the orders_products table is a wasted bit of data, at least to me.

in addition, having a MPN is for help in SELLING a product.  i would argue if you are using it for keeping track of inventory, having the same product_id that has different MPN numbers is another recipe for an inventory disaster.

perhaps others feel differently; but i think this thought process is why there are 3 different addresses carried over in the orders table.  which to me is another waste of storage.

**Comment by scottcwilson** [comment]
> MPNs are assigned by the manufacturer, and therefore rarely change for a specific product. 

From experience, this is sadly untrue.  And you want to preserve historic information. 

> i think this thought process is why there are 3 different addresses carried over in the orders table. which to me is another waste of storage.

Storage is cheap, and billing, shipping and customer addresses are often different.  

"I don't need it" doesn't mean others don't.

**Comment by torvista** [comment]
I have instances where a product gets updated and has a new MPN, but the product is outwardly unchanged..so only the MPN changes in the product data/I don't create a new product.
Regarding the order_products table, perhaps the MPN could be suffixed to the product name to maintain the history? It would not affect a search on the name,

**Comment by drbyte** [comment]
> > Remember that you may want the `orders_products` table to also store the information as it was at the time of an order.
> 
> i disagree. MPNs are assigned by the manufacturer, and therefore rarely change for a specific product. and if there is a change, in theory it is a different product. ensuring that you have the MPN at the time of the order in the orders_products table is a wasted bit of data, at least to me.
> 
> in addition, having a MPN is for help in SELLING a product. i would argue if you are using it for keeping track of inventory, having the same product_id that has different MPN numbers is another recipe for an inventory disaster.

My point was merely in allowing one to either reconstruct an order correctly, if needed, or to research and understand what an order was comprised of at the time the order was placed. 
A storeowner may need to query order history regarding inventory details. 
If an MPN is not captured in the order details, then you'll never be able to query its existence if it were merely part of an old product that has since been deleted by the storeowner, having been replaced by something else. 

> perhaps the MPN could be suffixed to the product name to maintain the history

Perhaps. But then something like Edit Orders which reconstructs the order so that returns/substitutions/coupons/discounts/taxes are handled and recalculated, might not know how best to handle the MPN suffix. If it is present separately then it's easy to retain/review or even utilize.

**Comment by torvista** [comment]
>If it is present separately then it's easy to retain/review or even utilize.

I'm always a bit reticent to proffer my model-centric view of the world. 
But if MPN goes in the product table, it should go where relevant as it's own entity such as this.
While the idea for having it is based on selling/demands from Google, it could also be used in a purchasing plugin for making an order to the supplier and in providing a pricelist to dealers for those who do wholesale pricing.

**Comment by scottcwilson** [comment]
> [MPN] could also be used in a purchasing plugin for making an order to the supplier and in providing a pricelist to dealers for those who do wholesale pricing.

That's certainly the dream!

**Comment by proseLA** [comment]
> My point was merely in allowing one to either reconstruct an order correctly, if needed, or to research and understand what an order was comprised of at the time the order was placed. A storeowner may need to query order history regarding inventory details. If an MPN is not captured in the order details, then you'll never be able to query its existence if it were merely part of an old product that has since been deleted by the storeowner, having been replaced by something else.

what about creating a change log table for the MPN?  that table could store the product_id, old_mpn, new_mpn and the date_changed?  in those situations where a store owner needs to reconstruct what the MPN was, the data in fact would be there to do such a task.

we would then not be over-populating the orders_products table with unnecessary fields for every order.

if we got that route, might i add that any new date fields just use a unix timestamp to store the date?  in my experience, mysql date fields are always a bit problematic, and storing them as an `int` is far easier when needing to format it to the user in their preferred format.

> Storage is cheap, and billing, shipping and customer addresses are often different.
> 
> "I don't need it" doesn't mean others don't.

storage may be cheap, but that does not take into account the real cost.  at some point, individuals may realize making things more efficient is far better in the long run.

in addition, i never said i don't need it.  i know more about UPC/EAN/MPNs etc than i ever care to admit.  i said i do not agree with storing it in the orders_product table.  i am all for putting it into the products table.  or better yet just use the change log table for knowing what the MPN is/was at any point in time.  then it does not even need to reside in the products table as well.

**Comment by torvista** [comment]
Since the bigger picture is UPC/EAN are also demanded by Google in addition to MPN perhaps these should be borne in mind at this point although the EAN / ISBN / UPC international code required is product type dependant, and so not common to all shops/should be a customisation?
Personally I've added MPN, EAN and google category number to my products table, there are enough observers available so this can be done without code modifications.

I made the original suggestion as I believe MPN is required for all products?

**Comment by lat9** [comment]
Er, a 32-byte varchar is a significant database growth for an existing site that has **_2,000,000+_** ordered products!

**Comment by scottcwilson** [comment]
> Er, a 32-byte varchar is a significant database growth for an existing site that has 2,000,000+ ordered products!

The default will be null for those existing ordered products.  This takes an additional bit but let's round that up to a byte.
2,000,000 extra bytes is 1.9M.   I have clients that post larger PDFs than this.

**Comment by scottcwilson** [comment]
I don't really have a dog in this fight because I would just add these fields for clients that need them.  But if you don't capture the data at order time in the orders_products table, it really doesn't make sense to add it to the products table, because the order-time data will be lost as products are deleted, modified, updated through external scripts, etc. 

I think enough has been said on this topic.  @drbyte will decide and we'll move forward.

**Comment by proseLA** [comment]
> I am confused about why your'e so animated about this topic. 

scott, again try and stick to the issue without the personal characterizations.  i would appreciate it.

> > [MPN] could also be used in a purchasing plugin for making an order to the supplier and in providing a pricelist to dealers for those who do wholesale pricing.
> 
> That's certainly the dream!

> But if you don't capture the data at order time in the orders_products table, **it really doesn't make sense to add it to the products table,** because the order-time data will be lost as products are deleted, modified, updated through external scripts, etc.

i thoroughly disagree.  

how could one add the MPN to a pricelist if it is not included in the products table?

this is an attribute that is used so that customers can find a product _prior_ to the sale of said product.  as @torvista stated in his original comment, it is needed for any [google product feed](https://support.google.com/merchants/answer/6324482?hl=en).  the feed is a marketing tool for store owners to further market their products, allowing potential customers to do a google search for a MPN and find the store.

**Comment by drbyte** [comment]
While Zen Cart needs to be aware of product-centric information that can be fed to 3rd-party traffic-generation/marketing tools (like Google) and fulfillment systems (like Shipwire/Shipstation, even broader ERP systems), it is not a full-blown inventory-management platform.  

IMO having an "items" (naming open to suggestions) table (or small set of relational tables) where MPN/UPC/EAN can be associated with products and product-attribute-combinations, might be the best way to provide an intermediary way to store the necessary identifiers for the sake of product-feeds (and some incoming traffic, where customers might actually search for UPC/MPN/etc) and for storeowner administration.

That said ...

@proseLA you've objected to a lot of things in the above discussion, but I'm more interested in a proposal of what is the best way in your opinion to add MPN/UPC/EAN to core. What's your proposal?

For the record, I'm opposed to modeling all of it as a changelog. That's a whole separate rabbit trail that will lead us into tripling the database size because people will want to track changelogs for not only MPN but also all product/category/attribute details and customer details and address book details and ... well, nearly everything. Let someone add that with a plugin that creates database stored-procedures that do all the changelog tracking. Entirely outside the scope of the topic at hand.

**Comment by proseLA** [comment]
> @proseLA you've objected to a lot of things in the above discussion, but I'm more interested in a proposal of what is the best way in your opinion to add MPN/UPC/EAN to core. What's your proposal?
> 
in reviewing this thread, this is what I have objected to:

- the storage of an MPN in the orders_products table.
- the idea that storage is cheap.
- the personal characterizations.
- the idea that if the MPN is not stored in the orders_products table, then it makes no sense to store it in the products table.

for those people who wanted the ability to reconstruct the MPN at the time of the order, I recommended having a change log.  (I did not see the need for this sort of ability, the same way I do not need to know what the customer address was at the time of an order; but I wanted to offer a solution to those that do.)

I am of the opinion that your single table that contains MPN/EAN/UPC is the best method for handling this new addition.  and that this table needs to be keyed off of product/attribute combination.

and while there, it would only make sense to throw in an inventory quantity number.

while you have stated that zen-cart is not an inventory management system, every client that I have worked with uses it as such.  I know that woo-commerce includes the ability to keep track of inventory based on every product attribute combination.  I think ZC should do this out of the box else one needs to go to the plugins to do it or create new product_ids for every product attribute combination in order that one can keep track of inventory.

**Comment by drbyte** [comment]
> while you have stated that zen-cart is not an inventory management system, every client that I have worked with uses it as such

My sincerest apologies to all of them ... because doing inventory mgmt from within the Zen Cart admin sucks!

**Comment by torvista** [comment]
>doing inventory mgmt from within the Zen Cart admin sucks!

I would have thought this is only possible for online-only businesses.
Those with telephone sales, email sales, in-store sales like us, would have to use some ERP/invoicing/stock control software tailored to their business that would update ZC stock and prices on demand, as we do (push a csv).
ZC is just another sales channel but absolutely not adequate as a ERP, and I don't think it should go there either.
